util/metrics.py

```diff
@@ -0,0 +1,280 @@
+samples respectively.
+
+See --help to see further details.
+
+Code apapted from https://github.com/bioinf-jku/TTUR to use PyTorch instead
+of Tensorflow
+
+Copyright 2018 Institute of Bioinformatics, JKU Linz
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+import os
+import pathlib
+
+import numpy as np
+import torch
+
+from scipy import linalg
+from torch.nn.functional import adaptive_avg_pool2d, interpolate
+import random
+
+
+try:
+    from tqdm import tqdm
+except ImportError:
+    # If not tqdm is not available, provide a mock version of it
+    def tqdm(x):
+        return x
+
+
+def get_activations(
+    dataloader,
+    model,
+    domain,
+    batch_size,
+    dims,
+    device,
+    nb_max_img,
+):
+    """Calculates the activations of the pool_3 layer for all images.
+
+    Params:
+    -- dataloader  : JG-like dataloader (it can be a list of tensors too)
+    -- model       : Instance of inception model
+    -- domain      : image domain
+    -- batch_size  : Batch size of images for the model to process at once.
+                     Make sure that the number of samples is a multiple of
+                     the batch size, otherwise some samples are ignored. This
+                     behavior is retained to match the original FID score
+                     implementation.
+    -- dims        : Dimensionality of features returned by Inception
+    -- device      : Device to run calculations
+    -- nb_max_img  : number max of images used for activations compute
+
+    Returns:
+    -- A numpy array of dimension (num images, dims) that contains the
+       activations of the given tensor when feeding inception with the
+       query tensor.
+    """
+    model.eval()
+
+    if batch_size > len(dataloader):
+        print(
+            (
+                "Warning: batch size is bigger than the data size. "
+                "Setting batch size to data size"
+            )
+        )
+        batch_size = len(dataloader)
+
+    pred_arr = np.empty((len(dataloader), dims))
+
+    start_idx = 0
+
+    if nb_max_img < len(dataloader):
+        if isinstance(dataloader, list):
+            dataloader = dataloader[:nb_max_img]
+        else:
+            index = random.sample(np.range(nb_max_img))
+            subset = torch.utils.data.Subset(dataloader, index)
+            dataloader = subset
+
+    # print("batchsize", batch_size)
+
+    for batch in tqdm(dataloader, total=len(dataloader) // batch_size):
```

Comment:
`dataloader` returns a batch of size `train_batch_size` which can be different than `batch_size`